### PR TITLE
ensure language radio buttons are selected correctly

### LINKED ISSUE #3036: when setting language for default value in template creation, radio button is not getting highlighted
on DEV, using template completely:new

field Example 1 has a default value, but no language specified

I edit the template to specify the language of the default value, and in the language selection box, the top radio button for "Select language" never gets highlighted when I search for and then select a language.

### PR DESCRIPTION
**Pull request by peetucket**
## Why was this change made?

Fixes #3036 -- set the radio buttons correctly based on the language selection

## How was this change tested?

Localhost

### CODE REVIEW

File: src/components/editor/inputs/InputLang.jsx
Comment (peetucket) [inline src/components/editor/inputs/InputLang.jsx:23]: this lets us keep track of the selected radio button separately from the existence of the `lang` value and makes the component code below easier to understand ... the default value is based on the existence of a language selection coming in though

File: src/components/editor/inputs/InputLang.jsx
Comment (peetucket) [inline src/components/editor/inputs/InputLang.jsx:40]: this method is called when a user picks a language in the type ahead compontent: 

1. we use `null` to indicate no language is selected, `lang` can never be undefined (see line 20 above)
2. set the currently selected radio button appropriately

File: src/components/editor/inputs/InputLang.jsx
Comment (peetucket) [inline src/components/editor/inputs/InputLang.jsx:46]: this function is called when a user clicks one of the radio buttons:

1. record the button checked in state
2. if they selected "no language", we will clear the language in state

File: src/components/editor/inputs/InputLang.jsx
Comment (peetucket) [inline src/components/editor/inputs/InputLang.jsx:78]: 1. we now control if the radio button is checked or not based on the current value in state
2. our new function to handle when the user clicks the radio button

File: src/components/editor/inputs/InputLang.jsx
Comment (peetucket) [inline src/components/editor/inputs/InputLang.jsx:106]: 1. we now control if the radio button is checked or not based on the current value in state
2. our new function to handle when the user clicks the radio button

File: src/components/editor/inputs/InputLang.jsx
Comment (peetucket) [inline src/components/editor/inputs/InputLang.jsx:102]: 1. get rid of the `name="lang"` to prevent the browser from trying to set the value of the radio button based on the grouping (we set based on our value in state)
2. i don't think we need an `id` or `data-testid` here, just get rid of for simplicity

File: src/components/editor/inputs/InputLang.jsx
Comment (peetucket) [inline src/components/editor/inputs/InputLang.jsx:70]: get rid of the `name="lang"` to prevent the browser from trying to set the value of the radio button based on the grouping (we set based on our value in state)

File: src/components/editor/inputs/InputLang.jsx
Comment (peetucket) [inline src/components/editor/inputs/InputLang.jsx:99]: get rid of the name="lang" to prevent the browser from trying to set the value of the radio button based on the grouping (we set based on our value in state)

File: src/components/editor/inputs/InputLang.jsx
Comment (peetucket) [inline src/components/editor/inputs/InputLang.jsx:58]: Previously the user could click the "Select language" radio button but then not choose any language before clicking `Submit`.  This gives them an error if they try to do that.

File: src/components/editor/inputs/InputLang.jsx
Comment (justinlittman) [inline src/components/editor/inputs/InputLang.jsx:55]: No alerts. We should disable the Submit button instead when in bad state.

**Review by justinlittman — CHANGES_REQUESTED** [review]

File: src/components/editor/inputs/InputLang.jsx
Comment (peetucket) [reply src/components/editor/inputs/InputLang.jsx:55]: OK, though this may get pretty tricky to detect the bad state (I think you'd need to be constantly checking as the user types in the text box).  May not be worth the trouble if it hasn't come up before.

Comment (peetucket) [reply src/components/editor/inputs/InputLang.jsx]: actually came up with an alternate way of doing this (disable submit button and show an inline alert), required some extra state though

File: src/components/editor/inputs/InputLang.jsx
Comment (peetucket) [inline src/components/editor/inputs/InputLang.jsx:124]: this alert will be conditionally shown if the user tries to submit the form in a bad state (we also disable the submit button until they correct the issue)

**Comment by peetucket** [comment]
Alert triggered if you select the "language" radio button but don't put anything in there

Screen Shot 2021-09-29 at 3 46 55 PM

File: src/components/editor/inputs/InputLang.jsx
Comment (ndushay) [reply src/components/editor/inputs/InputLang.jsx:44]: Maybe this could be a comment for the function??  it's pretty intricate.

File: src/components/editor/inputs/InputLang.jsx
Comment (ndushay) [reply src/components/editor/inputs/InputLang.jsx:50]: ditto on making it a comment for the function

File: src/components/editor/inputs/InputLang.jsx
Comment (ndushay) [reply src/components/editor/inputs/InputLang.jsx:62]: excellent.

File: src/components/editor/inputs/InputLang.jsx
Comment (ndushay) [reply src/components/editor/inputs/InputLang.jsx:70]: also excellent.

File: src/components/editor/inputs/InputLang.jsx
Comment (ndushay) [reply src/components/editor/inputs/InputLang.jsx:82]: MUCH clearer

File: src/components/editor/inputs/InputLang.jsx
Comment (ndushay) [reply src/components/editor/inputs/InputLang.jsx:99]: ah!  also nicer.

File: src/components/editor/inputs/InputLang.jsx
Comment (ndushay) [reply src/components/editor/inputs/InputLang.jsx:111]: much much clearer

File: src/components/editor/inputs/InputLang.jsx
Comment (ndushay) [reply src/components/editor/inputs/InputLang.jsx:24]: this makes sense.  I'm wondering if there is a better name for radioButtonValue so it can be clear as a boolean.  `languageIsSelected` ?  `langIsSelectedRadioButton` ?  `langRadioSelector` ?   dunno if it's worth it, but "present" and "absent" are unusual values for a boolean switch.

**Review by ndushay — COMMENTED** [review]
Given how hard it was to figure this out ... it makes me want to have tests.  I'll leave it to @justinlittman to decide if tests are worth it.

One suggestion you can take or leave.

File: src/components/editor/inputs/InputLang.jsx
Comment (peetucket) [reply src/components/editor/inputs/InputLang.jsx:24]: yeah, present and absent are unusual choices, maybe renaming those values to something like "langSelected" (instead of "present") and "noLangSelected" (instead of "absent") will make it clear what is happening -- note this state value is not a boolean, it's the string value of the radio button we will use below to decide which one to check, which is why I called it "radioButtonValue"

Comment (peetucket) [reply src/components/editor/inputs/InputLang.jsx:27]: created constants for those radio button values and then used the better named constants through the code, also makes it more DRY and avoids a case where a typo in a string radio button value that is sprinkled throughout causes things to break

File: src/components/editor/inputs/InputLang.jsx
Comment (peetucket) [reply src/components/editor/inputs/InputLang.jsx:51]: added inline function comments

File: src/components/editor/inputs/InputLang.jsx
Comment (peetucket) [inline src/components/editor/inputs/InputLang.jsx:141]: allow us to disable the submit button if the user is in a bad state, this will also trigger the alert above to show

**Comment by peetucket** [comment]
There is one edge case not dealt with elegantly ... since the modal persists even after you close it, you can do this:

1. Pick a field with "no language specified" and click to get the modal.
2. Click the "Select a language" radio button.
2. Pick a language, e.g. type "En" and then select "English".
3. Click Submit.
4. Open the same modal again by selecting "English" next to the same field.
5. Click the "No language selected" radio button.  Note that the type ahead language field is disabled but the value still shows "English" (confusing, but not fatal).
6. Click Submit.
7. The language is now back to unspecified .. this is good and everything worked.
8. But...now open the same language modal again and note that even though the "No language selected" radio button is correctly pre-selected, the "English" language is still shown in the disabled type ahead field.  
9. Now click the "Select language" radio button and then hit "Submit" without doing anything else.  You'll get the alert claiming you need to select a language.  A bit confusing.

I think this stems from the fact that the state correctly has the `lang` value set in all cases, but the type ahead component value is there from the previous entry.  This is a pretty convoluted path and doesn't cause data errors, just a bit confusing for the user, so not sure if we want to spend any time on it or not.

File: src/components/editor/inputs/InputLang.jsx
Comment (justinlittman) [inline src/components/editor/inputs/InputLang.jsx:31]: I'd prefer to not use an alert here. It is not consistent with the rest of the app. The Submit button should be disabled when the form is in an invalid state (i.e., when the user has selected to provide a language, but has not done so.).

**Review by justinlittman — CHANGES_REQUESTED** [review]
Yeah, so this does need a feature test.

Here is a relevant existing test: https://github.com/LD4P/sinopia_editor/blob/main/__tests__/feature/editing/editingLiteral.test.js#L124-L161

For testing more variations, I would create a separate feature test based on ^^.

File: src/components/editor/inputs/InputLang.jsx
Comment (peetucket) [reply src/components/editor/inputs/InputLang.jsx:31]: Got it, I'll ditch the alerts but leave the submit button enabling/disabling.

**Comment by peetucket** [comment]
There were two existing tests that I just added some expectations to (looking at the radio buttons that were selected or not to verify).  Do you think this is sufficient?

File: __tests__/feature/editing/editingLiteral.test.js
Comment (justinlittman) [inline __tests__/feature/editing/editingLiteral.test.js:147]: You shouldn't need `waitFor` here. Try `const radioButtons = await screen.findAllByRole("radio")`

**Review by justinlittman — APPROVED** [review]
Approve pending consideration of comment.

File: __tests__/feature/editing/editingLiteral.test.js
Comment (peetucket) [reply __tests__/feature/editing/editingLiteral.test.js]: you are correct, i switched to just `await` and it worked

**Pull request merged**
